### PR TITLE
Commit Rule

### PR DESCRIPTION
**Pull request by lazynina**

### CODE REVIEW

**Comment by lazynina** [comment]
Current dependencies on/for this PR:
* feature/pos-process-block
  * **PR #727** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/727" target="_blank">Graphite</a> 
    * **PR #728** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/728" target="_blank">Graphite</a> 
      * **PR #729** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/729" target="_blank">Graphite</a> 
        * **PR #730** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/730" target="_blank">Graphite</a> 
          * **PR #732** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/732" target="_blank">Graphite</a> 
            * **PR #733** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/733" target="_blank">Graphite</a> 
              * **PR #734** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/734" target="_blank">Graphite</a> 
                * **PR #735** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/735" target="_blank">Graphite</a> 
                  * **PR #737** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/737" target="_blank">Graphite</a> 
                    * **PR #739** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/739" target="_blank">Graphite</a>  👈
                  * **PR #736** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/736" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/739?utm_source=stack-comment).

**Comment by lazynina** [comment]
Current dependencies on/for this PR:
* feature/pos-process-block
  * **PR #727** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/727" target="_blank">Graphite</a> 
    * **PR #728** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/728" target="_blank">Graphite</a> 
      * **PR #729** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/729" target="_blank">Graphite</a> 
        * **PR #730** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/730" target="_blank">Graphite</a> 
          * **PR #732** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/732" target="_blank">Graphite</a> 
            * **PR #733** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/733" target="_blank">Graphite</a> 
              * **PR #734** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/734" target="_blank">Graphite</a> 
                * **PR #735** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/735" target="_blank">Graphite</a> 
                  * **PR #737** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/737" target="_blank">Graphite</a> 
                    * **PR #739** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/739" target="_blank">Graphite</a>  👈
                  * **PR #736** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/736" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/739?utm_source=stack-comment).

File: lib/pos_blockchain.go [hunk 1]
```diff
@@ -481,9 +482,164 @@ func (bc *Blockchain) pruneUncommittedBlocks(desoBlock *MsgDeSoBlock) error {
 // commitGrandparents commits the grandparent of the block if possible.
 // Specifically, this updates the CommittedBlockStatus of its grandparent
 // and flushes the view after connecting the grandparent block to the DB.
-func (bc *Blockchain) commitGrandparents(desoBlock *MsgDeSoBlock) error {
-	// TODO: Implement me.
-	return errors.New("IMPLEMENT ME")
+func (bc *Blockchain) commitGrandparents() error {
```
Comment (tholonious) [inline lib/pos_blockchain.go:485]: consider a more self-descriptive function name that describes exactly what the function does
```suggestion
func (bc *Blockchain) runCommitRuleOnBestChain() error {
```

File: lib/pos_blockchain.go [hunk 2]
```diff
@@ -481,9 +482,164 @@ func (bc *Blockchain) pruneUncommittedBlocks(desoBlock *MsgDeSoBlock) error {
...
+			return nil
+		}
+		if !canCommit {
+			var exists bool
+			currentBlock, exists = bc.bestChainMap[*currentBlock.Header.PrevBlockHash]
+			if !exists {
+				return errors.Errorf("commitGrandparents: Block %v not found in best chain map", currentBlock.Hash.String())
+			}
+		}
+	}
```
Comment (tholonious) [inline lib/pos_blockchain.go:505]: Why do we need to loop over the ancestors of the current block here? If I understand this loop correctly, it's intended to find the highest block that can be committed.

It should be fine to call `bc.canCommitGrandparent(currentBlock)` on just the current tip. If the grandparent can be committed, then you can run the commit rule on the grandparent. It shouldn't be possible for the grandparent to not be committable, while another ancestor is committable. 
```suggestion
	blockToCommit, canCommit, _ := bc.canCommitGrandparent(currentBlock)
	if !canCommit {
		return nil
	}	
```

File: lib/pos_blockchain.go [hunk 3]
```diff
@@ -481,9 +482,164 @@ func (bc *Blockchain) pruneUncommittedBlocks(desoBlock *MsgDeSoBlock) error {
...
+	return nil
+}
+
+// canCommitGrandparent determines if the grandparent of the current block can be committed.
+// The grandparent can be committed if there exists a direct parent-child relationship
+// between the grandparent and parent of the new block, meaning the grandparent and parent
+// are proposed in consecutive views, and the "parent" is an ancestor of the incoming block (not necessarily consecutive views).
+// Additionally, the grandparent must not already be committed.
+func (bc *Blockchain) canCommitGrandparent(currentBlock *BlockNode) (_grandparentBlockHash *BlockHash, _canCommit bool, _committedBlockSeen bool) {
```
Comment (tholonious) [inline lib/pos_blockchain.go:533]: What is the _committedBlockSeen return param used for? IMO, it's sufficient to just use the _canCommit param to define whether the grand can or cannot be committed.

File: lib/pos_blockchain.go (hunk 2 above)
Comment (lazynina) [reply lib/pos_blockchain.go:505]: Great! okay I was unsure of this assumption.

File: lib/pos_blockchain.go (hunk 3 above)
Comment (lazynina) [reply lib/pos_blockchain.go:533]: This was only necessary when I was looping over ancestors. If we hit a "committed" block then we could stop the search. This is no longer necessary and will be removed.

**Review by tholonious — APPROVED** [review]

**Comment by lazynina** [comment]
### Merge activity

* **Oct 19, 3:48 PM**: @lazynina started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/739).

**Pull request merged**